tomviz/SaveLoadTemplateReaction.cxx

```diff
@@ -0,0 +1,121 @@
+/* This source file is part of the Tomviz project, https://tomviz.org/.
+   It is released under the 3-Clause BSD License, see "LICENSE". */
+
+#include "SaveLoadTemplateReaction.h"
+
+#include "ActiveObjects.h"
+#include "ModuleManager.h"
+#include "Pipeline.h"
+#include "RecentFilesMenu.h"
+#include "Utilities.h"
+
+#include <QApplication>
+#include <QDebug>
+#include <QInputDialog>
+#include <QJsonArray>
+#include <QJsonDocument>
+#include <QJsonObject>
+#include <QString>
+
+namespace tomviz {
+
+SaveLoadTemplateReaction::SaveLoadTemplateReaction(QAction* parentObject, bool load, QString filename)
+  : pqReaction(parentObject), m_load(load), m_filename(filename)
+{}
+
+void SaveLoadTemplateReaction::onTriggered()
+{
+  if (m_load) {
+    loadTemplate(m_filename);
+  } else {
+    saveTemplate();
+  }
+}
+
+bool SaveLoadTemplateReaction::saveTemplate()
+{
+  bool ok;
+  QString text = QInputDialog::getText(tomviz::mainWidget(), tr("Save Pipeline Template"),
+                                         tr("Template Name:"), QLineEdit::Normal, QString(), &ok);
+  QString fileName = text.replace(" ", "_");
+  if (ok && !text.isEmpty()) {
+    QString path = QApplication::applicationDirPath() +
+                    "/../share/tomviz/templates/" + fileName + ".tvsm";
```

Comment:
In an installed version of Tomviz the application is unlikely to be able to write to this path. I think the template generated by the user whould be stored in the "Tomviz directory". This location is currently used for custom operators, see `MainWindow`. This is the code that defines the path:

```CPP
QStringList locations =
     QStandardPaths::standardLocations(QStandardPaths::HomeLocation);
    QString home = locations[0];
    QString path = QString("%1%2tomviz").arg(home).arg(QDir::separator());
```

We should move this code to a utility function and then use it in both places. We can have a sub directory called `templates`.

Reply:
This has been moved into a utility function and is being used in `importCustomTransform` as well as for saving and loading template files.